c_src/device/cairo.c

```diff
@@ -0,0 +1,173 @@
+#include <cairo.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/fb.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "cairo_ctx.h"
+#include "comms.h"
+#include "device.h"
+#include "fontstash.h"
+#include "ops/script_ops.h"
+
+const char* device = "/dev/fb0";
+
+typedef struct {
+  int fd;
+
+  struct fb_var_screeninfo var;
+  struct fb_fix_screeninfo fix;
+} cairo_fb_t;
+
+cairo_fb_t g_cairo_fb = {0};
+
+extern device_info_t g_device_info;
+extern device_opts_t g_opts;
+
+int device_init(const device_opts_t* p_opts,
+                device_info_t* p_info,
+                driver_data_t* p_data)
+{
+  if (g_opts.debug_mode) {
+    log_info("cairo %s", __func__);
+  }
+
+  if ((g_cairo_fb.fd = open(device, O_RDWR)) == -1) {
+    log_error("Failed to open device %s: %s", device, strerror(errno));
+    return -1;
+  }
+
+  if (ioctl(g_cairo_fb.fd, FBIOGET_VSCREENINFO, &g_cairo_fb.var)) {
+    log_error("Failed to get fb_var_screeninfo: %s", strerror(errno));
+    return -1;
+  }
+
+  if (ioctl(g_cairo_fb.fd, FBIOGET_FSCREENINFO, &g_cairo_fb.fix)) {
+    log_error("Failed to get fb_fix_screeninfo: %s", strerror(errno));
+    return -1;
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)calloc(1, sizeof(CAIROcontext_t));
```

Comment:
You might not need the cast here. There are other places in this PR where you are `malloc`ing and not casting. Both `calloc` and `malloc` return `void *`.